### ISSUE TITLE
Store GW: Stream chunk fetching; Reduce memory used per query.

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#3318]
Following @pracucci idea from last Thanos Community Sync in order to reduce and predict better memory usage of Thanos Store Gateway per query.

The idea is to move Block index lookup to a separate, first stage and do it concurrently for all blocks that potentially can match the query. We pack all in some kind of merge sort series set. 

Then we can have a defined number of workers that get series IDs and fetches chunks for them send to stream (in the right order).

This way we can highly control the overall resources per requests used. This is even more important because we overfetch chunk bytes from object storage to avoid API rate limits. Holding those bytes as short as possible and only for few series / blocks, make a lot of sense. (: 

Help wanted, whoever have time first. It's big refactor.

**Comment by bwplotka** [comment]
Some comments we made on code while 1:1 to @yeya24

**Comment by kakkoyun** [comment]
@bwplotka @yeya24 I think we're missing the content of the comments? Could you please check the link again?

**Comment by bwplotka** [comment]
Just click the link and scroll down you see comments in the code added

**Comment by bwplotka** [comment]
@yeya24 is on it :hugs:

**Comment by bwplotka** [comment]
Still very viable

**Comment by GiedriusS** [comment]
There is a huge optimization possibility here (https://github.com/thanos-io/thanos/pull/4271).

**Comment by GiedriusS** [comment]
The problem here is that we do sorting before sending so it's much harder than it seems :/

**Comment by GiedriusS** [comment]
@fpetkovski isn't this something similar to what you have implemented with block series client?

**Comment by fpetkovski** [comment]
Yes, we already do something like this. Chunks are fetched in batches and sent via gRPC. What we could improve on is to add some pooling so that bytes for sent chunks get recycled. I am not sure how this would work with the async nature of gRPC though, maybe we should move to planetscale/vtproto first?

**Comment by GiedriusS** [comment]
That wouldn't help as it is a problem with grpc-go itself. Maybe that could be a point in favor of switching to something like https://github.com/storj/drpc. But that's a totally separate thing so maybe we can close this?

**Comment by fpetkovski** [comment]
From what I understand, Mimir does some pooling in store gateway, but I don't know the exact details: https://grafana.com/blog/2023/07/06/breaking-the-memory-barrier-how-grafana-mimirs-store-gateway-overcame-out-of-memory-errors. 

I agree it can be treated as a separate issue though.
